### PR TITLE
[EC-83] UInt256 / Int256

### PR DESCRIPTION
**Pull request by whysoserious**

### CODE REVIEW

File: src/main/scala/io/iohk/ethereum/vm/Int256.scala [hunk 1]
```diff
@@ -0,0 +1,201 @@
+package io.iohk.ethereum.vm
```
Comment (rtkaczyk) [inline src/main/scala/io/iohk/ethereum/vm/Int256.scala]: Can we move it to `util`?

Comment (mcsherrylabs) [inline src/main/scala/io/iohk/ethereum/vm/Int256.scala]: As discussed Radek, leave it in the VM for now.

Comment (rtkaczyk) [inline src/main/scala/io/iohk/ethereum/vm/Int256.scala]: Putting it in a different package doesn't mean we'll be using it elsewhere. With the package structure as is we'll have a cyclic dependency between `domain` and `vm`. Just saying.

File: src/main/scala/io/iohk/ethereum/vm/Int256.scala [hunk 2]
```diff
@@ -0,0 +1,201 @@
...
+
+  def apply(bytes: IndexedSeq[Byte]): UInt256 = {
+    require(bytes.length <= Size, s"Input byte array cannot be longer than $Size: ${bytes.length}")
+    UInt256(bytes.foldLeft(BigInt(0)) { (n, b) => (n << 8) + (b & 0xff) })
+  }
+
+  def apply(int256: Int256): UInt256 = new UInt256(toUnsignedBigInt(int256.n))
+
+  implicit def numericToUInt256[N: Integral](n: N): UInt256 =
+    apply(BigInt(implicitly[Integral[N]].toLong(n)))
```
Comment (rtkaczyk) [inline src/main/scala/io/iohk/ethereum/vm/Int256.scala]: We probably don't want to convert BigInts automatically as that can conversion can be lossy

File: src/main/scala/io/iohk/ethereum/vm/Int256.scala [hunk 3]
```diff
@@ -0,0 +1,201 @@
...
+
+  val Size: Int = 32
+
+  val Zeros: ByteString = ByteString(Array.fill[Byte](Size)(0))
+
+  val Modulus: BigInt = BigInt(2).pow(Size * 8)
+
+}
+
+trait Int256Like[T <: Int256Like[T]] extends ScalaNumber with Ordered[T] with ScalaNumericConversions {
```
Comment (rtkaczyk) [inline src/main/scala/io/iohk/ethereum/vm/Int256.scala]: What do we need `ScalaNumericConversions` for?

**Comment by ntallar** [comment]
Apart from @rtkaczyk comments, LGTM!

File: src/main/scala/io/iohk/ethereum/vm/Int256.scala (hunk 1 above)
Comment (rtkaczyk) [inline src/main/scala/io/iohk/ethereum/vm/Int256.scala]: @mcsherrylabs @whysoserious on the branch for EC-85, that @AlanVerbner is working on, `DataWord` escapes the `vm` package and is used in transaction execution. That's one more reason to put it in `util` package.

File: src/main/scala/io/iohk/ethereum/vm/Int256.scala (hunk 2 above)
Comment (whysoserious) [inline src/main/scala/io/iohk/ethereum/vm/Int256.scala]: Yes, but let's fix that in next iteration.

File: src/main/scala/io/iohk/ethereum/vm/Int256.scala (hunk 3 above)
Comment (whysoserious) [inline src/main/scala/io/iohk/ethereum/vm/Int256.scala]: I just used conventions used in other numeric types.

**Pull request merged**